lib/vdsm/virt/livemerge.py

```diff
@@ -420,6 +421,11 @@ def _start_commit(self, drive, job):
             raise exception.MergeFailed(str(e), job=job.id)
 
     def _base_needs_extend(self, drive, job, base_info):
+        """
+        Return True, required_size if base need to be extended, otherwise
+        False, None. The required size is the current size to pass for
+        VM.extend_volume().
+        """
```

Comment:
With `namedtuples` and `dataclasses` available, I am not a big fan of returning simple tuples, but I understand it is a matter of taste and here it is fine as it is.

However, an alternative to using tuples altogether might be to use the returned size to signify both IF it needs to be extended and by how much:
- Return 0 if base does not need to be extended
- Return required_size (>0) if base needs to be extended

Reply:
Since it returns either (True, NUMBER,) or (False, None,), the first argument is redundant.

Reply:
> Since it returns either (True, NUMBER,) or (False, None,), the first return value is redundant.

Reply:
`None` vs `int` as @mz-pdm mentioned is another alternative then. But at least I learnt the detail of the single chunk extension for 0 size :)

Reply:
Hofefully you can unlearn this quicky, next version will eliminate this quirk.

Reply:
Inventing a namedtuple for _base_needs_extend just to avoid using tuple unpacking
does not make sense, and it create worse code:

    extend = self._base_needs_extend()
    if extend.needed:
        self._vm.extend_volume(drive, drive.volumeID, extend.new_size)

This can make sense if the namedtuple is used in many places, like VolumeSize
and BlockInfo.

Returning None is even worse, and need to guess what is the meaning of None:

    new_size = self._base_needs_extend()
    if new_size is not None:
        self._vm.extend_volume(drive, drive.volumeID, extend.new_size)

Now the caller have to ask, "what does None new_size mean?"

I kept the extend_needed, new_size result.

Reply:
It doesn't smell well when something returns two values where one is immediately determined by the other one (why e.g. dict.get() doesn't use this pattern?).  But it's storage code and if storage wants it this way, I don't mind.  Maybe / Some / Optional (etc., people apparently desire to be creative and not to miss the chance to invent new names for the same thing) is used for similar purposes in strongly typed or supposedly strongly typed languages; I don't know whether something similar is used in Python, probably not as it would make little sense there.

Reply:
In Python there is type hinting [Optional](https://docs.python.org/3/library/typing.html#typing.Optional), which is a hint to programmers, but it does not enforce the type. It can do type checks with `mypy` (integrated in `flake8`), but you need to run it to get the type mismatches.

Anyway, type hinting is not used anywhere in the project, and I guess to start using it requires consensus. I think we can keep this as it is right for now and maybe change or update the design separately if we think it is worth.

Reply:
> It doesn't smell well when something returns two values where one is immediately determined by the other one (why e.g. dict.get() doesn't use this pattern?).  But it's storage code and if storage wants it this way, I don't mind.  Maybe / Option / Optional (etc., people apparently desire to be creative and not to miss the chance to invent new names for the same thing) is used for similar purposes in strongly typed or supposedly strongly typed languages; I don't know whether something similar is used in Python, probably not as it would make little sense there.

Reply:
Thanks for the pointer.  So the Python type system still assumes the standard "value or None" pattern.

Reply:
In this case we want to return 2 values:
- if we need to extend
- the new size

Sure we can be clever and determine if we need to extend by returning a special value
(e.g. None), but I think that clever code is worse, and using the language features
(multiple return value, tuple unpacking) and create more readable code is the right way.

Reply:
IMHO the problem comes because the function tries to do both things at the same time, ideally the design would have two functions for that. One returning a `bool`, the other an `int`.

However, in this case it is difficult to separate them since you discover both things at the same time.

I'm sure we can think of a nicer design, but with the options that we are discussing, there is no big 'right' or 'wrong' alternative.